v2/robotmk/src/scheduling/cleanup.rs

```diff
@@ -0,0 +1,121 @@
+    for suite in suites {
+        debug!(
+            "Cleaning up working directory {} of suite {}",
+            suite.working_directory, suite.name
+        );
+        let _ = cleanup_working_directory(
+            &suite.working_directory,
+            &suite.working_directory_cleanup_config,
+        )
+        .context(format!(
+            "Error while cleaning up working directory of suite {}",
+            suite.name
+        ))
+        .map_err(log_and_return_error);
+    }
+}
+
+fn cleanup_working_directory(
+    directory: &Utf8Path,
+    cleanup_config: &WorkingDirectoryCleanupConfig,
+) -> Result<()> {
+    let dir_entries = directory
+        .read_dir_utf8()?
+        .filter_map(|dir_entry_result| {
+            dir_entry_result
+                .context(format!("Failed to list entry of directory {}", directory))
+                .map_err(log_and_return_error)
+                .ok()
+        })
+        .collect::<Vec<_>>();
+    for dir_entry in match cleanup_config {
+        WorkingDirectoryCleanupConfig::MaxAgeSecs(max_age_secs) => {
+            dir_entries_to_remove_max_age(dir_entries, *max_age_secs)
+        }
+        WorkingDirectoryCleanupConfig::MaxExecutions(max_executions) => {
+            dir_entries_to_remove_max_executions(dir_entries, *max_executions)
+        }
+    } {
+        let _ = remove_dir_entry(&dir_entry).map_err(log_and_return_error);
+    }
+    Ok(())
+}
+
+fn dir_entries_to_remove_max_age(
+    dir_entries: Vec<Utf8DirEntry>,
+    max_age_secs: u64,
+) -> Vec<Utf8DirEntry> {
+    let now = SystemTime::now();
+    dir_entries
+        .into_iter()
+        .filter(|dir_entry| {
+            is_dir_entry_too_old(dir_entry, max_age_secs, &now)
+                .context("Failure while checking if directory entry should be removed")
+                .map_err(log_and_return_error)
+                .unwrap_or(false)
+        })
+        .collect()
+}
+
+fn is_dir_entry_too_old(
+    dir_entry: &Utf8DirEntry,
+    max_age_secs: u64,
+    now: &SystemTime,
+) -> Result<bool> {
+    match now.duration_since(
+        dir_entry
+            .metadata()
+            .context(format!(
+                "Failed to retrieve metadata of {}",
+                dir_entry.path()
+            ))?
+            .modified()
+            .context(format!(
+                "Failed to retrieve modification time of {}",
+                dir_entry.path()
+            ))?,
+    ) {
+        Ok(duration) => Ok(duration.as_secs() > max_age_secs),
+        // now is earlier than modification time --> directory was modified in the meantime, so keep
+        Err(_) => Ok(false),
+    }
+}
+
+fn dir_entries_to_remove_max_executions(
+    mut dir_entries: Vec<Utf8DirEntry>,
+    max_executions: usize,
+) -> Vec<Utf8DirEntry> {
+    if max_executions > dir_entries.len() {
+        return vec![];
+    }
+    dir_entries.sort_by_key(|dir_entry| dir_entry.path().to_path_buf());
```

Comment:
They should be sorted by age, no?

Reply:
I rely on the timestamps in the working directory names for the sorting by age. I added a comment for this. Would you prefer an explicit sorting by age?

Reply:
This seems like unnecessary coupling to me. Moreover, I'm not convinced, that the two implementations behave the same in all edge cases.